Normalise default initial size behaviour across all OS platforms

citkane opened the pull request:
In MacOS, if the user does not explicitly call `webview_set_size`, the window runs but fails to render in the GUI.
This behaviour differs from Linux and Windows, and this can cause confusion for the end user.

Edit: Also resolves #1248 and #1251 for Windows, where the default size causes a visual glitch when the user also sets a custom size.

This PR addresses the ~issue~ issues by ~setting and calling~ dispatching a default window render size from the backend at ~Main window creation~ class initialisation.

citkane commented:
@SteffenL , I don't think that this build "should" be failing?
The failure is on a Windows test run, which is unrelated to the changed file.

citkane commented:
I updated this PR to take into consideration #1251
This now normalise default window sizing behaviour across all OS platforms, and does not set a default size if it has already been explicitly set by the user.

rcorre commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -412,9 +412,6 @@ public:
       m_dpi = get_window_dpi(m_window);
-      constexpr const int initial_width = 640;
-      constexpr const int initial_height = 480;
-      set_size(initial_width, initial_height, WEBVIEW_HINT_NONE);
```
This is not enough to prevent the "jump" on windows. The initial size is still being set here: https://github.com/webview/webview/blob/e2bd2e55c5984e43ebb3cc92fe1ebcb727917cb2/core/include/webview/detail/backends/win32_edge.hh#L407-L408

citkane replied:
Mmmmm...
Looks like MS requires values for some of those: https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-createwindoww

Strange though, should be `0` width `0`height, ie. not there...

I am not in a Windows dev env ATM, perhaps you could fiddle around with the parameters and report back if there is a magic formula that prevents the glitch?

Currently `CreateWindowW` is called in the class constructor, so we can't pass user values for width and height. Moving it out of the class constructor is probably going to pull on a loose thread that we don't want to pull.

@SteffenL @phildremi Any insights?

rcorre replied:
> Strange though, should be 0 width 0height, ie. not there...

It still renders the toolbar, so you briefly see a toolbar with no window "body"

>  perhaps you could fiddle around with the parameters and report back if there is a magic formula that prevents the glitch

The only reliable thing I've found is to set the initial position size in `CreateWindowW`. So we'd either need to defer that to `run` (which, as you noted, could potentially be a breaking change), or have a way to pass the position/size to the `webview` constructor.

Or, as a hacky approach, initially pass some very large negative values for x/y so the window renders offscreen until `set_size` is called.

citkane replied:
@rcorre I invited you to be a collaborator on my Webview fork. You should have full access to the `mac-default-size` branch (this PR).

citkane replied:
@rcorre I updated this PR by moving `ShowWindow`, `UpdateWindow` and `SetFocus` to the end of `set_size_impl` (with a backstop) in `win32_edge_engine`. CI tests are passing, but would you please confirm if this visually works and resolves the glitch?

rcorre replied:
Thanks @citkane! Your latest patch works for me. I missed that webview was calling `ShowWindow` in the constructor. Moving that to `run` means I can resize/reposition the window without any jumps.

citkane commented:
@SteffenL @phildremi Would you please give PR this a review when you get chance.

Resolves: #1251 as well as the [initial PR purpose](https://github.com/webview/webview/pull/1249#issue-2869844650)

citkane reviewed (commented):
Confirmed to work: https://github.com/webview/webview/pull/1249/files#r1971782515

rcorre commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -644,6 +635,11 @@ protected:
+      UpdateWindow(m_window);
+      SetFocus(m_window);
+    }
     return {};
```
Commented on the other thread, but yes, this works. Thank you!

SteffenL commented on core/include/webview/detail/engine_base.hh:
```diff
@@ -315,6 +322,9 @@ protected:
       }
     }
   }
+  bool m_is_size_set = false;
```
Please make these variables private, and make the `const` ones `static const` or `static constexpr`.

citkane replied:
Done in latest commit

SteffenL commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -644,6 +635,11 @@ protected:
                    SWP_NOZORDER | SWP_NOACTIVATE | SWP_NOMOVE |
                        SWP_FRAMECHANGED);
     }
+    if (m_owns_window && !m_is_size_set) {
```
Can you please explain why we only show/update the window if `!m_is_size_set`?

citkane replied:
By including the `!m_is_size_set` backstop, we preserving the existing logic where the [existing affected block](https://github.com/webview/webview/blob/e2bd2e55c5984e43ebb3cc92fe1ebcb727917cb2/core/include/webview/detail/backends/win32_edge.hh#L514-L518) was only ever called **once** in the `win32_edge_engine` class constructor (called with the default size, thus causing the visual glitch when the user sets size).

The patched block behaviour thus remains that `ShowWindow`, `UpdateWindow` and `SetFocus` are only ever called **once** on `m_window`:
- When the user calls `webview_set_size` before `webview_run`
- else with the default size when the user calls `webview_run`

Calls to `webview_set_size` <ins>after</ins> `webview_run` still work as expected as illustrated in this [minimal reproduction](https://gist.github.com/citkane/6162e1fd345aa7f180fa0b9349fd07d2).

I don't understand _why_ it works like this, but it is safer if we preserve the existing logic.

SteffenL commented:
I agree the library should have the same behavior across platforms when a size isn't explicitly set by the user, but more importantly it shouldn't break on macOS when a size isn't set explicitly.

Ideally we should allow setting an initial size when constructing the `webview` class along with other things (without piling on constructor parameters). I worked on that among other things in PR #766 but it was never merged and it hasn't really been worked on since then. See `create_options_builder`, `webview_create_options_t` in that PR for how that was done.

SteffenL commented:
Come to think of it, the library doesn't require the user to call `run()` if they manage their own event loop.

citkane commented:
> Come to think of it, the library doesn't require the user to call `run()` if they manage their own event loop.

mmm...
That will change existing behaviour for Windows users managing their own event loop. This is however (unavoidably I think) in the nature of stopping the visual glitch and standardising behaviour across OS platforms. Short of your #766 which could inject user size into `webview_create`, I feel that the approach here in this PR is direct and simple for _most_ use cases. Maybe it could be wrapped in a minor release describing the behaviour change case?

SteffenL commented:
I tried to experiment a bit with other solutions on macOS such as by setting an initial window size, but there's a problem where the window becomes visible with the default size before it's resized with `set_size()`, similar to the glitch that happens on Windows.

In my opinion we should really stop displaying the window by default during construction, and instead allow changing the visibility using something like `set_visible(true)`. Calling `set_size()` should still be optional.

This should in my opinion be the desired behavior, but I hesitate to suddenly break backward compatibility.

There's a solution for providing backward compatibility in PR #766 but that approach can lead to complex code. See `webview_create_options_t::minimum_required_version`:

```
  // Minimum required library version. This can be used to change the available
  // features and behaviors within the library. Set this to either
  // WEBVIEW_VERSION to use the latest features and behaviors provided by the
  // library, an older version for selective backward-compatibility, or zero
  // ({0, 0, 0}) to use the library's minimum supported version.
  // Note that the minimum supported version will be bumped when old and
  // deprecated features eventually are removed from the library.
```

SteffenL commented:
Expanding on my previous comment, here's a rough implementation without backward compatibility.

https://github.com/webview/webview/compare/master...SteffenL:webview:impl-set-visibility

SteffenL commented:
Thank you for the summary and other ideas.

I may have another idea for this PR. Would it be doable to use `dispatch()` during construction instead of changing `run()`, `set_size()`, etc? That way you can schedule code to run when the event loop begins. Do however keep in mind that on Windows, `embed()` spins the even loop temporarily so `dispatch()` would have to be used afterward.

As for PR #766, I've cherry-picked code from it over time so eventually when there isn't anything useful left there then we should probably just close it.

SteffenL commented:
There are some other places that will temporarily spin the event loop so the `dispatch()` idea takes a little bit more work. I've updated my branch but the changes aren't ideal.

I do however believe that it behaves mostly as you would expect it to:

* Sets a default window size on all platforms.
* Displays the window by default which retains backward compatibility.
* `set_size()` is optional which fixes the inconsistency on macOS.
* Delays showing the window on Windows until the event loop begins, aligning behavior with Linux and macOS.
* `set_visible()` needs a separate discussion but I left it in there for now. If called then the default behavior is skipped.

I'll try to see if I can avoid modifying `run()` a bit later. A silly idea from the top of my head might be to re-`dispatch()` until the right time.

citkane commented:
Great idea @SteffenL to use `webview_dispatch`. The behaviour of the latest commit to this PR should now be:
| Issue | Linux | Windows | Mac | OS Normalised
|:-----------------|:-----------------|:-----------------|:-----------------|:-----------------:|
|  Visual jump after `webview_set_size` called  | N/A  | Fixed | N/A  |Yes |
|  No-show window if `webview_set_size` not called  | N/A  | N/A   | Fixed   | Yes |
|  No-show window if `webview_set_size` not called with custom event-loop (ie. `webview_run` not called)| Unchanged - should show  | Unchanged/Fixed - should show without visual glitch   | Fixed - should now show   | Yes | 

- `set_size_default` is a new protected method of `engine_base`, and it in turn calls `dispatch(set_size)`. This new method is called at the absolute end of each platform's class constructor.
- The `m_is_size_set` backstop prevents `set_size_default` from actioning if the user has already called `webview_set_size`.
- `webview_run` is thus NOT required to be called to set a default size.

This _should_ work for custom event loops (ie. `run` is never called), but I have not explicitly tested it in this scenario as it is not a rabbit hole I wish to go down... Perhaps you could confirm if the behaviour is as expected?

It works in all other scenarios on all platforms using [this minimal reproduction](https://gist.github.com/citkane/6162e1fd345aa7f180fa0b9349fd07d2).

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -234,22 +235,28 @@ protected:
-      objc::msg_send<void>(
-          m_window, "setFrame:display:animate:"_sel,
-          CGRectMake(rect.origin.x, rect.origin.y, width, height), YES, NO);
+      dispatch(fn);
```
This change is needed to call `webview_set_size` from a child thread on MacOS via `dispatch`. See the [minimal reproduction](https://gist.github.com/citkane/6162e1fd345aa7f180fa0b9349fd07d2), which crashes without this patch.

The implementation here is a bit explicit, but I feel that it is warranted:
- We are better off not impacting the existing function behaviour when run on the `main` thread, ie. a direct synchronous call.
- Other Webview API calls that a user may use from a child thread are potentially also buggy, and this patch establishes a workable/flexible pattern.

SteffenL commented:
Can you please check whether your solution still works as expected on Windows when using `bind()`? I came to think of the fact that it also spins the event loop temporarily so callables passed to `dispatch()` may end up being called during a call to `bind()`. See call to `deplete_run_loop_event_queue()` in `win32_edge::add_user_script_impl()`.

citkane commented:
Too right, it does indeed trigger the default size prematurely...

See the updated minimal reproduction for the test, and the latest commit for the resolution.

citkane commented on core/include/webview/detail/engine_base.hh:
```diff
@@ -161,8 +163,13 @@ protected:
+        m_user_scripts.end(), add_user_script_impl(js)));
+    m_is_size_set = size_set_state;
+    return script;
   }
```
Here we prevent premature run-loop triggering of the default window size.

citkane replied:
See the latest commit for a working implementation where we put the default sizing back onto the event queue if needed.

Ideally we should only implement this in `win32_edge`, but it depends on the private flag member `m_is_size_set` in `engine_base`.

citkane commented:
This is not so simple as the temporary run loop cancels out the initial default size from the queue...

citkane commented:
The latest commit to this PR resolves the outstanding win32 event loop issue in a sane and clean way.
All issues should now be fixed, @SteffenL if you would confirm if the custom run-loop scenario behaves as expected.

SteffenL commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -117,6 +117,7 @@ public:
         }
       }
     }
+    dispatch_size_default();
```
Since an existing window may be passed in with its desired default size, this may have unexpected behavior. A default size can be set with `initWithContentRect:styleMask:backing:defer:`. You can check my branch for clues. I forgot the exact details but I had some issues with the window not handling mouse clicks properly after sending the `center` message so I had to work around that.

SteffenL replied:
A simpler fix may be to just check whether we own the window before setting a default size.

SteffenL replied:
For all platforms, if an existing window is passed in then then we try not touch it more than necessary, and only provide a widget. Same goes with the application lifecycle, i.e. application setup, teardown, event loop, etc.

I have some code examples where this library has been integrated into existing GUI apps.

citkane replied:
See latest commit.

Would you please re-run the 3 timed out CI checks.

SteffenL replied:
Regarding the specific code in my branch, this only applies for macOS.

Here's the part that sets a default size when the window is created:

https://github.com/webview/webview/compare/master...SteffenL:webview:57df864ea94a71d860f9883620db37dae1a1bbc1#diff-5eaff9e0ea51897f3ace6f23bd53ee64eae06eeff8bf8a877a2937c507bdaa80R454

The following is just a minor tip in case you touch the same code as I did and get surprised with mouse click gliches.

Maybe the `center` message *must* be sent after `makeKeyAndOrderFront:`/`makeKeyWindow` and/or after `setContentView:`. I may just have gotten the order wrong the first time.

https://github.com/webview/webview/compare/master...SteffenL:webview:57df864ea94a71d860f9883620db37dae1a1bbc1#diff-5eaff9e0ea51897f3ace6f23bd53ee64eae06eeff8bf8a877a2937c507bdaa80R303

SteffenL replied:
Relevant code for default window size on Linux:

https://github.com/webview/webview/compare/master...SteffenL:webview:57df864ea94a71d860f9883620db37dae1a1bbc1#diff-5eab752f0992d19082647ff471504aa73995e086887cd5567bf59ad37fa34632R109

citkane replied:
Interesting, thanks for the links.

That is a broader refactoring that also touches on issues outside of this PR, and I do not disagree with them, but it will take me some time to absorb it.

For now, I would prefer to see this PR merged as it touches the existing code very lightly, addresses all of the stated issues and (please advise) does not disturb any existing behaviour (gremlins unrelated to this PR and all).

Moving forward, we could layer on the refactoring and also discuss:
- Standardising initial positioning. (MacOS seems to be the only one that centres the window)
- Making user window options more friendly, eg. xy size, xy position, border decorations, centering, etc. The current `WEBVIEW_HINT_` architecture is not easy to grasp as a user. Key will be to find what settings are common to all 3 platforms and present them to the user as a singular abstraction.
- Refactoring size/rendering issues into separate platform classes and distilling the existing main backend classes to be more concise/readable.

SteffenL commented on core/include/webview/detail/engine_base.hh:
```diff
@@ -134,7 +134,9 @@ window.__webview__.onUnbind(" +
-    return set_size_impl(width, height, hints);
+    set_size_impl(width, height, hints);
+    m_is_size_set = true;
+    return noresult{};
```
Please return the result of `set_size_impl()`.

citkane replied:
See latest commit.

Would you please re-run the 3 timed out CI checks.

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -650,6 +657,10 @@ private:
     return temp;
   }
+  static bool is_main_thread() {
+    mach_port_t pid = getpid();
```
Agreed that this would be a good topic for a separate discussion/PR.

I will update my minimal reproduction to use `dispatch` and re-set this code shortly.

Worth a mention here:
- Using `dispatch` for thread safety is undocumented, and using/not using it has inconsistent behaviour across OS platforms. Despite developer expectations, it has caused me much confusion and frustration so I consider it to be a serious user issue. I also do not believe that documentation is the solution to the issue. Consistent behaviour (fail or preferable succeed) across all platforms is.
- I thought `is_main_thread()` was working, but somehow missed that "not positive" does not mean that positive is ever being hit... My stupid bad. The principle is that, specific to Mac, the main thread id should equal the process id.

SteffenL commented:
It appears that the default window size isn't set on Linux when building the library against WebKitGTK API 6.0 and GTK 4. It works fine with WebKitGTK API 4.1 and GTK 3.

This happens because there's no built-in way in GTK 4 to resize a window after it has been set up. We can set the initial size as long as it's done early, and I guess it's too late here.

See `gtk_compat::window_set_size()`.

SteffenL commented:
Before this PR, the default window size with GTK 3 and 4 were the same. This PR fixes the default size when using GTK 3, and the issue persists with GTK 4.

In my branch it worked with GTK 4 because the default size was set early.

Would you like to look into that? The current state is an improvement regardless.

SteffenL commented:
While doing some testing on Debian 12.10 with the Xfce desktop environment, I noticed that both before and after this PR, the window will initially be displayed with a small size before being resized to the desired size.

I don't see that glitch in my branch.

Tested with WebKitGTK API 4.1/GTK 3 and WebKitGTK API 6.0/GTK 4.

SteffenL commented:
Solving all the issues found so far while retaining full backward compatibility has required some mental gymnastics, but I believe the code in the following branch resolves all the issues found so far.

https://github.com/webview/webview/compare/master...SteffenL:webview:impl-set-visibility-2

While this appears to work fine, I'm not 100% confident about the re-dispatching logic in `dispatch()` as the order in which dispatched functions are called could potentially be problematic.

citkane commented:
Thanks for pointing out the further issues with GTK. I have it resolved now. I do not have Xfce handy, so perhaps you could confirm if it resolved there too.

I caught another premature run-loop on MacOS, and that is also resolved.

Yes - some mental gymnastics indeed. Part of that is because each platform is handled in a different/not clear code flow structure with different semantics. I was trying to avoid refactoring, but eventually had to for sanity's sake. I have been checking in on your branch, but this is what has made sense to me on this branch:
```mermaid
flowchart TD
    n2["Class Constructor <br>"] --> n1["m_window_init()"]
    n1 --> n7["m_window_settings()"]
    n7 --> n8["dispatch_size_default"]
    n10["User maybe calls set_size?<br>"] --> n11["is_size_set=true <br>"] & n13["m_window_show <br>"]
    n8 --> n10 & n14["default size dispatch gets blocked if user size is set<br>"]
    n9(("Wait for run loop / event queue processing<br>")) --> n13
    n11 --> n14
    n14 --> n9

    n2@{ shape: rounded}
    n1@{ shape: text}
    n7@{ shape: text}
    n8@{ shape: diam}
    n10@{ shape: rect}
    n11@{ shape: tri}
    n13@{ shape: rect}
    n14@{ shape: text}
```

The flow and naming conventions shown in the chart are common for all 3 platforms in this PR. Further re-factoring down the line can start excluding all non-common methods into platform specific helper classes, and improve separation of concerns/readbility going forward.

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -84,39 +84,12 @@ private:
-          set_up_window();
-        }
-      }
-    }
```
This whole block has been moved into standardly named methods so that all platform class constructors become:
```
 m_window_init();
 m_window_settings(debug);
 dispatch_size_default(m_owns_window);
```

SteffenL replied:
The change seems reasonable to me.

I would however remove the `m_` prefix in the method names.

citkane replied:
Removed `_m` prefix in latest commit.

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -157,8 +130,7 @@ public:
-      auto app = get_shared_application();
-      objc::msg_send<void>(app, "setDelegate:"_sel, nullptr);
+      objc::msg_send<void>(m_app, "setDelegate:"_sel, nullptr);
       // Make sure to release the delegate we created.
```
I favoured initialising `m_app` in the class constructor instead of repeated calls to `get_shared_application`

SteffenL commented:
> I do not have Xfce handy, so perhaps you could confirm if it resolved there too.

With the latest commit I don't see the glitch anymore with Xfce.

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -250,7 +221,7 @@ protected:
 
-    return {};
+    return m_window_show();
   }
```
Standardised `m_window_show` semantic across all platforms and moved logic into there.

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -498,35 +469,7 @@ private:
-  }
-  void set_up_web_view() {
+  void m_window_settings(bool debug) {
     objc::autoreleasepool arp;
```
Normalised this across all platforms into the `m_window_init()` semantic.

citkane replied:
Normalised this across all platforms into the m_window_settings() semantic.

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -498,35 +469,7 @@ private:
-    if (m_owns_window) {
-      objc::msg_send<void>(m_window, "makeKeyAndOrderFront:"_sel, nullptr);
-    }
-  }
```
Normalised this across all platforms into the m_window_init() semantic.

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -498,35 +469,7 @@ private:
-    }
-  }
-  void set_up_web_view() {
+  void m_window_settings(bool debug) {
```
For now we just pass in `debug`, but as the code evolves this parameter scope can broaden.

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -650,11 +595,68 @@ private:
+    if (!m_is_window_shown) {
+      set_is_size_set(false);
+      dispatch_size_default(m_owns_window);
+    }
```
Another premature run-loop drain default size gremlin averted!

citkane replied:
This was actually unneeded because `m_window_init()` all completes before `dispatch_size_default`. It was actually introducing a double dispatch of the default size, so I removed it.

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -650,11 +595,68 @@ private:
     }
     return temp;
   }
+  void m_window_init() {
```
This does what the class constructor and `set_up_window` were doing, just DRYer and much less nesting.

SteffenL commented:
> I caught another premature run-loop on MacOS, and that is also resolved.

I forgot about that one too...

After a round of testing I haven't seen any odd behavior.

citkane commented:
As an aside, has there ever been any consideration/discussion about a common multi-platform embedding framework such as:
- [CEF](https://github.com/chromiumembedded/cef) (Chromium) or
- [Gecko](https://firefox-source-docs.mozilla.org/overview/gecko.html) (Firefox)

as an alternative to WebkitGtk/Webview2/WKWebView?

citkane commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -306,14 +306,204 @@ private:
+  noresult terminate_impl() override {
+    PostQuitMessage(0);
+    return {};
+  }
+  noresult dispatch_impl(dispatch_fn_t f) override {
+    PostMessageW(m_message_window, WM_APP, 0, (LPARAM) new dispatch_fn_t(f));
+    return {};
+  }
+
+  noresult set_title_impl(const std::string &title) override {
+    SetWindowTextW(m_window, widen_string(title).c_str());
+    return {};
+  }
+
+  noresult set_size_impl(int width, int height, webview_hint_t hints) override {
+    auto style = GetWindowLong(m_window, GWL_STYLE);
+    if (hints == WEBVIEW_HINT_FIXED) {
+      style &= ~(WS_THICKFRAME | WS_MAXIMIZEBOX);
+    } else {
+      style |= (WS_THICKFRAME | WS_MAXIMIZEBOX);
+    }
+    SetWindowLong(m_window, GWL_STYLE, style);
+
+    if (hints == WEBVIEW_HINT_MAX) {
+      m_maxsz.x = width;
+      m_maxsz.y = height;
+    } else if (hints == WEBVIEW_HINT_MIN) {
+      m_minsz.x = width;
+      m_minsz.y = height;
+    } else {
+      auto dpi = get_window_dpi(m_window);
+      m_dpi = dpi;
+      auto scaled_size =
+          scale_size(width, height, get_default_window_dpi(), dpi);
+      auto frame_size =
+          make_window_frame_size(m_window, scaled_size.cx, scaled_size.cy, dpi);
+      SetWindowPos(m_window, nullptr, 0, 0, frame_size.cx, frame_size.cy,
+                   SWP_NOZORDER | SWP_NOACTIVATE | SWP_NOMOVE |
+                       SWP_FRAMECHANGED);
+    }
```
Unchanged - moved code block

citkane commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -306,14 +306,204 @@ private:
+    auto wjs = widen_string(js);
+    std::wstring script_id;
+    bool done{};
+    webview2_user_script_added_handler handler{[&](HRESULT res, LPCWSTR id) {
+      if (SUCCEEDED(res)) {
+        script_id = id;
+      }
+      done = true;
+    }};
+    auto res =
+        m_webview->AddScriptToExecuteOnDocumentCreated(wjs.c_str(), &handler);
+    if (SUCCEEDED(res)) {
+      // Sadly we need to pump the even loop in order to get the script ID.
+      while (!done) {
+        deplete_run_loop_event_queue();
+      }
+    }
+    // TODO: There's a non-zero chance that we didn't get the script ID.
+    //       We need to convey the error somehow.
+    return user_script{
+        js, user_script::impl_ptr{new user_script::impl{script_id, wjs},
+                                  [](user_script::impl *p) { delete p; }}};
+  }
+
+  void
+  remove_all_user_scripts_impl(const std::list<user_script> &scripts) override {
+    for (const auto &script : scripts) {
+      const auto &id = script.get_impl().get_id();
+      m_webview->RemoveScriptToExecuteOnDocumentCreated(id.c_str());
+    }
+  }
+
+  bool are_user_scripts_equal_impl(const user_script &first,
+                                   const user_script &second) override {
+    const auto &first_id = first.get_impl().get_id();
+    const auto &second_id = second.get_impl().get_id();
+    return first_id == second_id;
+  }
+
+private:
```
Unchanged - moved code block

citkane commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -510,203 +696,23 @@ public:
+      UpdateWindow(m_window);
+      SetFocus(m_window);
+      m_is_window_shown = true;
     }
```
Unchanged - moved code block

citkane commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -510,203 +696,23 @@ public:
-    auto wjs = widen_string(js);
-    std::wstring script_id;
-    bool done{};
-    webview2_user_script_added_handler handler{[&](HRESULT res, LPCWSTR id) {
-      if (SUCCEEDED(res)) {
-        script_id = id;
-      }
-      done = true;
-    }};
-    auto res =
-        m_webview->AddScriptToExecuteOnDocumentCreated(wjs.c_str(), &handler);
-    if (SUCCEEDED(res)) {
-      // Sadly we need to pump the even loop in order to get the script ID.
-      while (!done) {
-        deplete_run_loop_event_queue();
-      }
-    }
-    // TODO: There's a non-zero chance that we didn't get the script ID.
-    //       We need to convey the error somehow.
-    return user_script{
-        js, user_script::impl_ptr{new user_script::impl{script_id, wjs},
-                                  [](user_script::impl *p) { delete p; }}};
-  }
-
-  void
-  remove_all_user_scripts_impl(const std::list<user_script> &scripts) override {
-    for (const auto &script : scripts) {
-      const auto &id = script.get_impl().get_id();
-      m_webview->RemoveScriptToExecuteOnDocumentCreated(id.c_str());
-    }
-  }
-
-  bool are_user_scripts_equal_impl(const user_script &first,
-                                   const user_script &second) override {
-    const auto &first_id = first.get_impl().get_id();
-    const auto &second_id = second.get_impl().get_id();
-    return first_id == second_id;
-  }
-
-private:
```
Unchanged - moved code block

SteffenL commented on core/include/webview/detail/engine_base.hh:
```diff
@@ -318,6 +319,19 @@ protected:
+      return;
+    };
+    dispatch([this]() {
+      if (!m_is_size_set) {
```
Note: On Windows, this lambda is executed 10 times when running the basic C++ example without calling `set_size()`, and 4 times with `set_size()`.

SteffenL commented on core/include/webview/detail/engine_base.hh:
```diff
@@ -318,6 +319,19 @@ protected:
     }
   }
 
+  void dispatch_size_default(bool m_owns_window) {
```
Note: In my Windows environment, when passing in an existing window, this function is called 7 times.

citkane replied:
See previous comment and latest PR. Should mitigate this.

citkane replied:
Also, `dispatch_size_default` will execute, but do nothing if `!m_owns_window`.

This helps keep code DRYer/cleaner from the calling side by avoiding repeated
```
if (m_owns_window) {
  dispatch_size_default();
};
```
conditional blocks.

From the calling side, it is implicit that setting the default size behaviour has something to do with `m_owns_window`, and in my opinion this is cleaner and more readable code.

SteffenL commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -859,6 +866,9 @@ private:
   // Blocks while depleting the run loop of events.
   void deplete_run_loop_event_queue() {
+    // prevent premature run loop triggering of the default window size
+    set_is_size_set(true);
```
I'm reading this as:

* The size has been set.
* Loop
* If window isn't visible already then the size has *not* been set anymore.

The flag is also set in `set_size()`, and it looks like the flag could be reset back to `false` unconditionally.

I understand the flag is used to prevent `set_size()` from being called during the temporary event loop but I wonder if some renaming would be doable to make this easier to read.

citkane replied:
See latest PR for improved semantics and additional backstop.

The scenario is that we are depleting the event queue outside of the main event loop, so we don't want to set a default size **yet** (but the lamda has **already** been dispatched in the class constructor). We thus need to deal with this "future". 
1) `default_size_backstop(true)` flags `m_is_size_set` to true so that the dispatched lamda will execute, but do nothing.
2) Loop. The lamda is removed from the event queue.
3) We evaluate:
[if the user has called `set_size` (`!m_is_window_shown`)]
and 
[if this isn't a destructor call (`!is_destructor`)] 
and then conditionally:
  - reset the backstop and 
  - put the `dispatch_size_default()` lamda back on the event queue.

citkane replied:
We _could_ ignore this like we do on MacOS (ie. `dispatch_size_default` is called **after** `window_init` in the class constructor), except that Windows also depletes the queue on `bind`, and I feel that this is semantically the correct block to apply the brakes.

I am open to an alternative implementation in `add_user_script_impl`, but that is already a rather dense method to consume.

citkane replied:
Flip flop, eat my hat...

I have implemented this in `add_user_script_impl` only and removed attendant complexities.
This should reduce unneeded execution of the `dispatch_size_default` inner lamda.

As an aside, in `add_user_script_imp`:
```
while (!done) {
  deplete_run_loop_event_queue();
}
```
is a loop within a loop, which was likely a cause of reciprocal execution of the size lamda (and should now no longer happen).

citkane replied:
See https://github.com/webview/webview/pull/1249/files#r2007965027

SteffenL commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -868,12 +878,18 @@ private:
   // CreateCoreWebView2EnvironmentWithOptions.
   // Source: https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/webview2-idl#createcorewebview2environmentwithoptions
   com_init_wrapper m_com_init;
+  void *window_ptr = nullptr;
```
I understand this variable is only used during construction so I would prefer a shorter lifespan. We already have `m_window`.

Maybe it could at least be renamed to something like `m_hwnd_or_phwnd` to make it more clear.

The passed-in window can be either a native window handle (HWND, which is a pointer), or, for backward compatibility reasons, `HWND*`.

citkane replied:
I take your point.
My eye was on standardised methods across all platforms, so I have simplified it across all by passing `*window` to `window_init(window)` and handling it further for each platform therein.

See latest commit.

SteffenL commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -628,12 +573,12 @@ private:
     objc::msg_send<void>(m_widget, "addSubview:"_sel, m_webview);
     objc::msg_send<void>(m_webview, "setFrame:"_sel,
                          objc::msg_send_stret<CGRect>(m_widget, "bounds"_sel));
+    objc::msg_send<void>(m_window, "setContentView:"_sel, m_widget);
```
This line should be part of the window setup, not the widget setup.

citkane replied:
Fixed in latest commit

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -616,6 +559,8 @@ private:
   return window.webkit.messageHandlers.__webview__.postMessage(message);\n\
 }");
+    set_up_widget();
+    objc::msg_send<void>(m_window, "makeKeyAndOrderFront:"_sel, nullptr);
```
Fixed in latest commit

citkane commented on core/include/webview/detail/backends/cocoa_webkit.hh:
```diff
@@ -616,6 +557,11 @@ private:
     add_init_script("function(message) {\n\
```
From here `bind` is indirectly executed in the class initialisation flow. This may have been spawning reciprocal executions of the `set_default_size` lamda, but this should now be reduced to one.

citkane replied:
I put this comment under the wrong backend, relates to win32_edge.
See https://github.com/webview/webview/pull/1249/files#r2009105779

SteffenL commented:
May I suggest using different wording for the "backstop"?

Maybe it's because I'm not into baseball, or maybe it's just due to my ignorance, but I'm not very familiar with that word.

What the flag should do is prevent/block the default window size from being set (prematurely) during the temporary event loop(s).

Some candidates from the top of my head:

* prevent default size
* (un)block default size
* running temp event loop
* acquire/release default size (b)lock - word "lock" may cause confusion

Instead of a boolean flag one could also use e.g. a counter:

```cpp
int counter{};
void acquire() noexcept { ++counter; }
void release() noexcept { --counter; }
bool is_engaged() const noexcept { return counter > 0; }
```

Then only handle e.g. default sizing depending on the return value of `is_engaged()`.

A simple variant with RAII:

```cpp
struct counter {
    struct auto_releasing {
        auto_releasing(counter* outer) noexcept : outer{outer} {}
        ~auto_releasing() { outer->release(); }
        counter* outer{};
    };
    auto_releasing acquire() noexcept { ++value; return auto_releasing{this}; }
    void release() noexcept { --value; }
    bool is_engaged() const noexcept { return value > 0; }
    int value{};
};
```

Usage:

```cpp
counter m_prevent_default_size_counter;
```

```cpp
{
    auto holder{m_prevent_default_size_counter.acquire()};
    // loop
} // holder.release()
```

Somewhere within in the loop this would take place:

```cpp
if (m_prevent_default_size_counter.is_engaged()) {
    // prevent default size
}
```

Something like this might be a little nicer but still not quite ideal. I did something similar in my branch above.

I'm not saying you have to do this, I'm just airing the idea. A simple flag should work fine the way it's currently used.

There could be cleaner ways but I'm not confident.

citkane commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -778,8 +793,8 @@ private:
                         "put_IsStatusBarEnabled failed"};
     }
     add_init_script("function(message) {\n\
```
From here bind is indirectly executed in the class initialisation flow. This may have been spawning reciprocal executions of the set_default_size lamda, but this should now be reduced to one.

citkane commented:
The common technical term that was eluding me is "[guard](https://en.wikipedia.org/wiki/Guard_(computer_science))" and not "[backstop](https://www.rte.ie/news/brexit/2019/0828/1071339-backstop/)".
The latest commit addresses this sematic, so apologies for the confusion caused.

I appreciate your concerns regarding the frequency and numeric discrepancies with which `dispatch_size_default` was being executed from win32_edge.

**The first way to look at the matter is**; "does the code work?", and for this we should measure how many times `set_size(m_initial_width, m_initial_height, WEBVIEW_HINT_NONE);` is being executed from inside the lamda (ie. "are the guards working?").

The result should be:
- 0 (user has set size) or
- 1 (user did not set size).

I believe that the answer was and remains with the latest commit; "_YES_, the code works". 
I will do some more explicit confirmation testing across platforms after this post.

**The second way to look at the matter is**; "is the code efficient and understandable?". I think that we will agree that the answer was; "_NO_", but hopefully this latest commit can change that to a; "_YES_".

There were a few factors confounding things for win32_edge that the latest and previous commits resolve:

1) `deplete_run_loop_event_queue` is called from the class destructor, so we were unwittingly executing `dispatch_size_default` during class destruction.  Mitigated in [27210bb](https://github.com/webview/webview/pull/1249/commits/27210bb4b561e219f21eb6c01c40e59db3955f72)

2) `add_init_script` is called during class initialisation, which indirectly executes `add_user_script_impl`, which in the [current master](https://github.com/webview/webview/blob/f1f42f0f164ec931ee05fae6e8bfaeba1c87b089/core/include/webview/detail/backends/win32_edge.hh#L671-L694) executes the `deplete_run_loop_event_queue` (which is a `while` loop) from within its own `while` loop, and unwanted strange behaviour ensues from looping inside of a loop...
In the latest commit I have:
 - refactored win32 `add_user_script_impl` and `deplete_run_loop_event_queue` to allow for not having to execute a loop inside of a loop.
 - added the `m_script_is_init` guard within `dispatch_size_default` so that the code execution flow is consistent across all 3 platforms (ie. dispatching the default size is firstly and explicitly done from the end of the class constructor). Windows will subsequently [guard it's execution and dispatch again](https://github.com/citkane/webview/blob/679b37bfa40afa79ecc158024f38f211ec2694cf/core/include/webview/detail/backends/win32_edge.hh#L484-L492) if the user has called `bind`.

SteffenL commented:
"Guard" sounds better than "backstop"!

Your change for `deplete_run_loop_event_queue()` conflicts with some work I did the other way so I've cleaned that up—hopefully without breaking something in the process—and submitted a PR for that specifically:

* #1265

Could you please submit a separate PR for the changes that involve `m_app`/`get_shared_application()` (macOS)?

citkane commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -309,6 +309,205 @@ private:
+      // conditionally add the default window size event back to the event queue
+      if (!m_is_window_shown) {
+        default_size_backstop(false);
+        dispatch_size_default(m_owns_window);
```
See latest guards and related discussions.

citkane commented:
> Could you please submit a separate PR for the changes that involve m_app/get_shared_application() (macOS)?

As requested in the latest commit.

See #1266 as a draft PR because it depends on this PR for CI tests to pass.

citkane commented on core/include/webview/detail/engine_base.hh:
```diff
@@ -318,6 +320,19 @@ protected:
 
+  void dispatch_size_default(bool m_owns_window) {
+    if (!m_owns_window || m_script_is_init) {
+      return;
```
`m_script_is_init` guards against Windows dispatching `set_size` prematurely in the execution of `add_init_script`. This is not relevant to other platforms, which do not empty the event queue during bind.

citkane commented:
FYI regarding https://github.com/webview/webview/pull/1249#issuecomment-2746255319

Given some logging:
```c++
  void dispatch_size_default(bool m_owns_window) {
    printf("dispatch_size_default\n");
    if (!m_owns_window || m_script_is_init) {
      return;
    };
    printf("dispatch_size_default: dispatching\n");
    dispatch([this]() {
      printf("dispatch_size_default: dispatched: execution\n");
      if (!m_is_size_set) {
        printf("dispatch_size_default: dispatched: execution: setting size\n");
        set_size(m_initial_width, m_initial_height, WEBVIEW_HINT_NONE);
      }
    });
  }
```
The output log pattern on Windows is now:
**Without user `set_size` and without user `bind`:**
```
dispatch_size_default
dispatch_size_default      
dispatch_size_default: dispatching
dispatch_size_default: dispatched: execution
dispatch_size_default: dispatched: execution: setting size
```
**Without user `set_size` and with user `bind`:**
```
dispatch_size_default
dispatch_size_default
dispatch_size_default: dispatching
dispatch_size_default: dispatched: execution
dispatch_size_default
dispatch_size_default: dispatching
dispatch_size_default: dispatched: execution
dispatch_size_default: dispatched: execution: setting size
```
**With user `set_size` and without user `bind`:**
```
dispatch_size_default
dispatch_size_default
dispatch_size_default: dispatching
dispatch_size_default: dispatched: execution
```

citkane commented:
@SteffenL Once you have merged #1265, I will split this PR into 5 discrete PR's:
- `engine_base` - prepare for backend changes
- x3 backend issue resolutions and refactors
- #1266 extra refactoring for `cocoa_webkit`

citkane commented on core/include/webview/detail/backends/win32_edge.hh:
```diff
@@ -778,8 +798,8 @@ private:
     add_init_script("function(message) {\n\
-  return window.chrome.webview.postMessage(message);\n\
-}");
+   return window.chrome.webview.postMessage(message);\n\
```
I have reverted this on the [Windows PR branch](https://github.com/citkane/webview/blob/0fbf7891eca777748d864baceac4b7cdffde8e66/core/include/webview/detail/backends/win32_edge.hh#L792-L794) that will supersede this PR, as respectively these also will:
- [Mac](https://github.com/citkane/webview/blob/refactor-cocoa_webkit2/core/include/webview/detail/backends/cocoa_webkit.hh)
- [Linux](https://github.com/citkane/webview/blob/refactor-gtk_webkitgtk/core/include/webview/detail/backends/gtk_webkitgtk.hh)

I will submit the PR's once #1267 and #1266 are merged (else CI will fail) and we can do final checks and tweaks within those.

citkane commented:
This PR has been split into:
- #1267
- #1268
- #1271 
- #1269 
- #1270
